Add remote cluster state documentation (#5726)
opensearch-project/documentation-website: _tuning-your-cluster/availability-and-recovery/remote-store/remote-cluster-state.md

---
layout: default
title: Remote Cluster State
nav_order: 5
parent: Remote-backed storage
grand_parent: Availability and recovery
---

# Remote Cluster State

Introduced 2.10
{: .label .label-purple }

Remote Cluster State protects against any cluster state metadata loss resulting due to quorum loss (permanently losing majority of cluster manager nodes) in the cluster.

> Cluster State in an internal data structure which contains the metadata of the cluster along with other information. 

Reviewer: `Cluster State in an internal data structure`  -> `Cluster State **is** an internal data structure`

> The metadata includes details about index metadata like settings, mappings, active copies of the shards, cluster level settings, aliases, templates and data streams. 

Reviewer 2: Combined this with line 16, so lets delete it.

```suggestion
```

> This metadata is managed by the elected cluster manager node and is essential for proper functioning of the cluster. When the cluster loses majority of the cluster manager nodes permanently, lets say 2 out of 3 cluster manager nodes are lost, then the cluster can experience data loss as there are no guarantees that latest cluster state metadata is present in the surviving nodes. Today, cluster durability is the function of the cluster manager node storage. And, persisting the state to remote provides better durability guarantees.

Reviewer 2: ```suggestion
Cluster durability is the function of the cluster manager node's storage. The cluster state metadata is managed by the elected cluster manager node and is essential for the cluster to properly function. When the cluster loses the majority of the cluster manager nodes permanently, then the cluster can experience data loss because the latest cluster state metadata might not be present in the surviving cluster manager nodes. Persisting the state of all the cluster manager nodes in the cluster to remote back-storage provides better durability.
```

> When remote cluster state feature is enabled, the cluster metadata will be published to a remote repository configured in the cluster. Note that, currently only index metadata will be persisted to remote store in OpenSearch 2.10.

Reviewer 2: ```suggestion
When the remote cluster state feature is enabled, the cluster metadata will be published to a remote repository configured in the cluster. As of OpenSearch 2.10 currently only index metadata will persist to remote-backed storage.
```

> When remote cluster state feature is enabled, the cluster metadata will be published to a remote repository configured in the cluster. Note that, currently only index metadata will be persisted to remote store in OpenSearch 2.10.

Reviewer 3: we should mention about global metadata persistence as well since in 2.11 we release that too

Author: Global metadata persistence is added in 2.x branch. These changes will get released in 2.12 version

> Any time, the new cluster manager nodes are launched after disaster recovery, they will bootstrap using the latest index metadata stored in the remote repository automatically. Consequently, the data of the indexes will also be restored when remote store is enabled.

Reviewer 2: ```suggestion
Any time new cluster manager nodes are launched after disaster recovery, the nodes will bootstrap using the latest index metadata stored in the remote repository automatically. Consequently, the data of the indexes will also be restored when the remote store is enabled.
```

> Any time, the new cluster manager nodes are launched after disaster recovery, they will bootstrap using the latest index metadata stored in the remote repository automatically. Consequently, the data of the indices will also be restored when remote store is enabled.

Reviewer 3: ```
Consequently, the data of the indices will also be restored when remote store is enabled.
```

this indicates remote state and store can be enabled separately. but as of now they cannot be. the cluster settings for remote state enabled/disabled is just a kill switch

Author: Should we add that in the limitation : 
```
Remote cluster state cannot be enabled without remote segment/translog being configured
```

> ## Configuring remote cluster state

Reviewer 2: ```suggestion
## Configuring the remote cluster state
```

> Remote cluster state settings are cluster level settings and can be enabled while bootstrapping the cluster. Once the remote cluster state is enabled, it can be disabled by updating the settings and performing a rolling restart of all the nodes.

Reviewer 2: ```suggestion
Remote cluster state settings can be enabled while bootstrapping the cluster. After the remote cluster state is enabled, it can be disabled by updating the settings and performing a rolling restart of all the nodes.
```

> To enable remote cluster state for a given cluster, provide the remote cluster state enabled setting and repository settings in `opensearch.yml` as show in the following example:

Reviewer 2: ```suggestion
To enable remote cluster state for a given cluster, add the following cluster-level and repository settings to the cluster's `opensearch.yml`:
```

```yml
# Enable Remote cluster state cluster setting
cluster.remote_store.state.enabled: true

# Remote cluster state repository settings
node.attr.remote_store.state.repository: my-remote-state-repo
node.attr.remote_store.repository.my-remote-state-repo.type: s3
node.attr.remote_store.repository.my-remote-state-repo.settings.bucket: <Bucket Name 3>
node.attr.remote_store.repository.my-remote-state-repo.settings.base_path: <Bucket Base Path 3>
node.attr.remote_store.repository.my-remote-state-repo.settings.region: <Bucket region>
```
{% include copy-curl.html %}

## Limitations

> 

Reviewer 2: The remote cluster state feature has the following limitations:

> - Currently, only index metadata is supported for upload and restore from remote store.

Reviewer 2: ```suggestion
- As of OpenSearch 2.10, only index metadata can uploaded and restored from remote-backed storage.
```

> - Currently, only index metadata is supported for upload and restore from remote store.

Reviewer 3: no longer a limitation after 2.11

Author: This limitation can be removed from 2.12

> - Unsafe bootstrap script cannot be run when remote cluster state is enabled. In case a majority of cluster-manager nodes are lost and the cluster goes down, the user needs to replace any remaining cluster-manager nodes and seed the nodes again for bootstrapping a new cluster.

Reviewer 2: Can you clarify what you mean by "unsafe"? How would a user verify that their bootstrap script is unsafe?

> - Remote cluster state cannot be enabled without remote store being configured

Reviewer 2: ```suggestion
- The remote cluster state cannot be enabled without first configuring remote-backed storage.
```

---
layout: default
title: Remote Cluster State
nav_order: 5
parent: Remote-backed storage
grand_parent: Availability and recovery
---

# Remote cluster state

Introduced 2.10
{: .label .label-purple }

> The _Remote Cluster State_ functionality for remote-backed storage, protects against any cluster state metadata loss resulting due to quorum loss, permanently losing the majority of cluster manager nodes, inside the cluster.

Reviewer 4: ```suggestion
The _remote cluster state_ functionality for remote-backed storage protects against any cluster state metadata loss resulting due to the permanent loss of the majority of cluster manager nodes inside the cluster.
```

> _Cluster State_ is an internal data structure which contains the metadata of the cluster, including: 

Reviewer 4: ```suggestion
_Cluster state_ is an internal data structure that contains the metadata of the cluster, including the following: 
```

- Index settings 
- Index mappings 
- Active copies of shards in the cluster 

> - Cluster level settings

Reviewer 4: ```suggestion
- Cluster-level settings
```

- Data streams
- Templates

> Cluster durability is the function of the cluster manager node's storage. The cluster state metadata is managed by the elected cluster manager node and is essential for the cluster to properly function. When the cluster loses the majority of the cluster manager nodes permanently, then the cluster can experience data loss because the latest cluster state metadata might not be present in the surviving cluster manager nodes. Persisting the state of all the cluster manager nodes in the cluster to remote back-storage provides better durability.

Reviewer 4: Add a line

> Cluster durability is the function of the cluster manager node's storage. The cluster state metadata is managed by the elected cluster manager node and is essential for the cluster to properly function. When the cluster loses the majority of the cluster manager nodes permanently, then the cluster can experience data loss because the latest cluster state metadata might not be present in the surviving cluster manager nodes. Persisting the state of all the cluster manager nodes in the cluster to remote back-storage provides better durability.

Reviewer 4: ```suggestion
The cluster state metadata is managed by the elected cluster manager node and is essential for the cluster to properly function. When the cluster loses the majority of the cluster manager nodes permanently, then the cluster may experience data loss because the latest cluster state metadata might not be present in the surviving cluster manager nodes. Persisting the state of all the cluster manager nodes in the cluster to remote-backed storage provides better durability.
```

> When the remote cluster state feature is enabled, the cluster metadata will be published to a remote repository configured in the cluster. As of OpenSearch 2.10 currently only index metadata will persist to remote-backed storage.

Reviewer 4: ```suggestion
When the remote cluster state feature is enabled, the cluster metadata will be published to a remote repository configured in the cluster. As of OpenSearch 2.10, only index metadata will persist to remote-backed storage.
```

> Any time new cluster manager nodes are launched after disaster recovery, the nodes will bootstrap using the latest index metadata stored in the remote repository automatically. Consequently, the data of the indexes will also be restored when the remote store is enabled.

Reviewer 4: ```suggestion
Any time new cluster manager nodes are launched after disaster recovery, the nodes will automatically bootstrap using the latest index metadata stored in the remote repository. Consequently, the index data will also be restored when the remote store is enabled.
```

## Configuring the remote cluster state

Remote cluster state settings can be enabled while bootstrapping the cluster. After the remote cluster state is enabled, it can be disabled by updating the settings and performing a rolling restart of all the nodes.

> To enable remote cluster state for a given cluster, add the following cluster-level and repository settings to the cluster's `opensearch.yml`:

Reviewer 4: ```suggestion
To enable the remote cluster state for a given cluster, add the following cluster-level and repository settings to the cluster's `opensearch.yml` file:
```

```yml
# Enable Remote cluster state cluster setting
cluster.remote_store.state.enabled: true

# Remote cluster state repository settings
node.attr.remote_store.state.repository: my-remote-state-repo
node.attr.remote_store.repository.my-remote-state-repo.type: s3
node.attr.remote_store.repository.my-remote-state-repo.settings.bucket: <Bucket Name 3>
node.attr.remote_store.repository.my-remote-state-repo.settings.base_path: <Bucket Base Path 3>
node.attr.remote_store.repository.my-remote-state-repo.settings.region: <Bucket region>
```
{% include copy-curl.html %}

## Limitations

> The remote cluster state feature has the following limitations:

Reviewer 4: ```suggestion
The remote cluster state functionality has the following limitations:
```

> - As of OpenSearch 2.10, only index metadata can uploaded and restored from remote-backed storage.

Reviewer 4: ```suggestion
- As of OpenSearch 2.10, only index metadata can be uploaded and restored from remote-backed storage.
```

> - Unsafe bootstrap script cannot be run when remote cluster state is enabled. In case a majority of cluster-manager nodes are lost and the cluster goes down, the user needs to replace any remaining cluster-manager nodes and seed the nodes again for bootstrapping a new cluster.

Reviewer 4: ```suggestion
- Unsafe bootstrap scripts cannot be run when the remote cluster state is enabled. When a majority of cluster-manager nodes are lost and the cluster goes down, the user needs to replace any remaining cluster manager nodes and reseed the nodes in order to bootstrap a new cluster.
```
